File: src/main/java/com/aarnolabs/haccs_ta2/phpexpgen/Dockerizer.java
package com.aarnolabs.haccs_ta2.phpexpgen;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.CopyOption;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

@Command(
    name = "dockerize",
    version = "dockerize 1.0",
    description = "Creates a docker image from an existing codebase and .cve.properties file",
    mixinStandardHelpOptions = true
)
public class Dockerizer extends BaseCmd {

    @Parameters(index = "0", description = "The path (directory) of the codebase to dockerize", paramLabel = "directory")
    private File pathToCodebase;

    @Option(names = {"-o", "--old"}, description = "If specified, uses the older PHP template")
    private boolean oldTemplate;

    public void copyFolder(Path src, Path dest, CopyOption copyOpt) throws IOException {
        try (Stream<Path> stream = Files.walk(src)) {
            stream.forEach(source -> copy(source, dest.resolve(src.relativize(source)), copyOpt));
        }
    }

    private void copy(Path source, Path dest, CopyOption copyOpt) {
        try {
            Files.copy(source, dest, copyOpt);
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    private void dockerize() throws IOException {
        File vulnFilePath = new File(pathToCodebase.getParentFile(), vulnFile.getPath());
        if (!vulnFilePath.exists()) {
            throw new RuntimeException("Vulnerable file does not exist: " + vulnFilePath);
        }

        File template = new File(
                oldTemplate ? "/home/jeikenberry/haccs/baseline2" : "/home/jeikenberry/haccs/baseline");
        if (!template.exists() || !template.isDirectory()) {
            throw new RuntimeException("Couldn't locate template directory: " + template);
        }

        File cveDir = new File(cve.toLowerCase());
        if (cveDir.exists()) {
            throw new RuntimeException("CVE directory already exists: " + cveDir);
        }

        File dataDir = new File(cveDir, "data");
        File workingCodebase = new File(dataDir, pathToCodebase.getName());
        System.out.printf("Copying %s template (from %s) to %s...\n", oldTemplate ? "old" : "new", template, cveDir);
        copyFolder(template.toPath(), cveDir.toPath(), LinkOption.NOFOLLOW_LINKS);
        copyFolder(pathToCodebase.toPath(), workingCodebase.toPath(), LinkOption.NOFOLLOW_LINKS);
        Files.copy(new File(".cve.properties").toPath(), new File(cveDir, ".cve.properties").toPath(),
                   StandardCopyOption.REPLACE_EXISTING);

        replaceLines(new File(cveDir, "README.md"), new int[] { 1, 3 },
                     new String[] { String.format("# %s", cve.toUpperCase()),
                             String.format("PHP exploit for %s (%s)", cve.toUpperCase(), vulnFile.getPath()) });
        replaceLines(new File(cveDir, "config.yml"), new int[] { 4, 16, 26 }, new String[] {
                "firmware_version: " + version, "exploit_name: " + cve.toUpperCase(), "cpe_product: " + cpeProduct });
        replaceLines(new File(cveDir, "data/doit.sh"), new int[] { 9, 10 }, new String[] {
                String.format("docker build -t aarno-%s . || exit_on_error \"Couldn't build docker container\"",
                              cve.toLowerCase()),
                String.format("docker run --rm --privileged -p 80:80 aarno-%s", cve.toLowerCase()) });
        String dockerLine = "COPY " + pathToCodebase.getName();
        if (oldTemplate) {
            dockerLine += "/ /var/www/html";
        } else {
            dockerLine += "/ /app";
        }
        replaceLines(new File(cveDir, "data/Dockerfile"), new int[] { 3 }, new String[] { dockerLine });
    }

    public static boolean isSorted(int[] a) {
        for (int i = 0; i < a.length - 1; i++) {
            if (a[i] >= a[i + 1]) {
                return false;
            }
        }
        return true;
    }

    private void replaceLines(File file, int[] lineNos, String[] toReplace) throws IOException {
        if (!isSorted(lineNos)) {
            throw new RuntimeException("Lines are not sorted: " + Arrays.toString(lineNos));
        }
        if (lineNos.length != toReplace.length) {
            throw new IllegalArgumentException(
                    String.format("Line numbers and corresponding string values do not line up: %d != %d",
                                  lineNos.length, toReplace.length));
        }
        List<String> lines = Files.readAllLines(file.toPath());
        for (int i = 0; i < lineNos.length; i++) {
            lines.set(lineNos[i] - 1, toReplace[i]);
        }
        try (PrintWriter out = new PrintWriter(file)) {
            for (String line : lines)
                out.println(line);
        }
    }

    @Override
    public Integer call() throws Exception {
        int result = super.call();
        if (result != 0)
            return result;

        if (!pathToCodebase.exists())
            throw new RuntimeException("Codebase does not exist: " + pathToCodebase);
        if (!pathToCodebase.isDirectory())
            throw new RuntimeException("Codebase is not a directory: " + pathToCodebase);

        dockerize();

        return 0;
    }

    // this example implements Callable, so parsing, error handling and handling user
    // requests for usage help or version help can be done with one line of code.
    public static void main(String... args) throws Exception {
        int exitCode = new CommandLine(new Dockerizer()).execute(args);
        System.exit(exitCode);
    }

}

File: sf/dockerizer.py
import configparser
import os
import shutil
import sys

from pathlib import Path

from .utils.file import replace_lines
from .utils.logging import ItemizedLogger


def dockerize(root_dir, path_to_codebase, log, use_old_template=False):
    old_cwd = os.getcwd()
    try:
        # To make things a bit easier let's chdir to root_dir, and we'll reset
        # this back later in the finally.
        log.debug(f"Setting CWD to '{root_dir}'")
        os.chdir(root_dir)

        # Read in the CVE ini file.
        config = configparser.ConfigParser()
        config.read('.cve.ini')
        cve_section = config['cve']

        # Extract the configuration variables needed for dockerization.
        cve = cve_section['id']
        cpe = cve_section['cpe']
        version = cve_section['version']
        vuln_file = cve_section['vuln_file']

        # Make sure the vulnerable file is where we are expecting it to be.
        if not Path(vuln_file).exists():
            log.error(
                f"ERROR: Vulnerable file '{vuln_file_path}' does not "
                'exist!'
            )
            return False

        # Determine which template we should dockerize from.
        if use_old_template:
            template_suffix = 'old'
        else:
            template_suffix = 'new'
        template_dir = '../../templates/baseline_' + template_suffix

        # Make sure the template directory can be located.
        if not Path(template_dir).is_dir():
            log.error(
                f"ERROR: Template directory '{template_dir}' either "
                'does not exist, or is not a directory!'
            )
            return False

        # Make sure that the directory for the CVE we will dockerize does not
        # already exist.
        cve_dir = cve.lower()
        if Path(cve_dir).exists():
            log.error(f"ERROR: CVE directory '{cve_dir}' already exists!")
            return False

        data_dir = os.path.join(cve_dir, 'data')
        working_codebase = os.path.join(data_dir, path_to_codebase)

        # Copy the template zygote into the CVE directory for which we will
        # dockerize.  Also, copy over the codebase into the <CVE>/data/
        # directory.
        log.info(
            f"Copying {template_suffix} template from '{template_dir}' "
            f"to '{cve_dir}'"
        )
        shutil.copytree(template_dir, cve_dir, symlinks=True)
        log.info(
            f"Copying codebase '{path_to_codebase}' into "
            f"'{working_codebase}'"
        )
        shutil.copytree(path_to_codebase, working_codebase, symlinks=True)

        # Copy over the cve.ini file.
        shutil.copyfile('.cve.ini', os.path.join(cve_dir, '.cve.ini'))

        # Replace all of the relevant lines in the template to be consistent
        # with what's in the CVE ini file.
        replace_lines(
            os.path.join(cve_dir, 'README.md'), [1, 3],
            ['# %s' % cve, 'PHP exploit for %s (%s)' % (cve, vuln_file)]
        )
        replace_lines(
            os.path.join(cve_dir, 'config.yml'), [4, 16, 26],
            [
                'firmware_version: ' + version,
                'exploit_name: ' + cve,
                'cpe_product: ' + cpe
            ]
        )
        doit_path = os.path.join(cve_dir, os.path.join('data', 'doit.sh'))
        replace_lines(
            doit_path, [9, 10],
            [
                'docker build -t aarno-%s . || exit_on_error "Couldn\'t build '
                'docker container"' % cve_dir,
                'docker run --rm --privileged -p 80:80 aarno-%s' % cve_dir
            ]
        )
        docker_line = 'COPY ' + os.path.basename(path_to_codebase)
        if use_old_template:
            docker_line += '/ /var/www/html'
        else:
            docker_line += '/ /app'
        dockerfile = os.path.join(cve_dir, os.path.join('data', 'Dockerfile'))
        replace_lines(dockerfile, [3], [docker_line])
    finally:
        # Set the CWD back to it's original path.
        log.debug(f"Resetting CWD back to '{old_cwd}'")
        os.chdir(old_cwd)

    # Success!
    return True
